# 03 - BatchNormalisation and Activations

After the input layer, values can be all over the place, really high or really low. Since tanh squashes everything between -1 and 1, extreme values get flattened and their gradients become basically zero. A zero gradient means we can never nudge those weights, so those neurons are dead and stop learning. This is the vanishing gradient problem.
Two fixes for this. First is Kaiming initialization, where we initialize the weights carefully so values dont explode or vanish from the start. Second is batch normalization, which normalizes values before they hit tanh so they stay in the sweet spot where gradients actually flow.

In [2]:
import torch
import torch.nn.functional as F
import pandas as pd
import random
import matplotlib.pyplot as plt

In [3]:
df = pd.read_parquet('data/0000.parquet')
comments = df['Comment'].dropna().tolist()
comments = [c.lower() for c in comments]

text = ''.join(comments)
chars = sorted(list(set(text)))
vocab_size = len(chars)

len(comments)

25430

In [4]:
stoi = {s:i+1 for i,s in enumerate(chars)}
stoi['|'] = 0
itos = {i:s for s,i in stoi.items()}    

In [5]:
random.shuffle(comments)

n1 = int(0.8 * len(comments))
n2 = int(0.9 * len(comments))

train = comments[:n1]
val = comments[n1:n2]
test = comments[n2:]

print(len(train))
print(len(val))
print(len(test))

20344
2543
2543


In [6]:
block_size = 3
def build_dataset(comments):
    X, Y = [], []
    for comment in comments: 
        context = [0] * block_size
        for ch in comment: 
            ix =stoi[ch]
            X.append(context)
            Y.append(ix)
            context = context[1:] + [ix]
        X.append(context)
        Y.append(0)
    
    X = torch.tensor(X)
    Y = torch.tensor(Y)
    return X,Y

In [7]:
Xtr, Ytr   = build_dataset(train)
Xval, Yval = build_dataset(val)
Xte, Yte   = build_dataset(test)

print(Xtr.shape, Ytr.shape)
print(Xval.shape, Yval.shape)
print(Xte.shape, Yte.shape)

torch.Size([3119222, 3]) torch.Size([3119222])
torch.Size([391022, 3]) torch.Size([391022])
torch.Size([393333, 3]) torch.Size([393333])


In [8]:
C  = torch.randn((74, 10))
W1 = torch.randn((30, 200))
b1 = torch.randn(200)
W2 = torch.randn((200, 74))
b2 = torch.randn(74)
parameters = [C, W1, b1, W2, b2]
for p in parameters:
    p.requires_grad = True

In [9]:
for i in range(200000):
    
    # minibatch
    ix = torch.randint(0, Xtr.shape[0], (32,))
    
    # forward pass
    emb = C[Xtr[ix]]
    h = torch.tanh(emb.view(-1, 30) @ W1 + b1)
    logits = h @ W2 + b2
    loss = F.cross_entropy(logits, Ytr[ix])
    
    # backward pass
    for p in parameters:
        p.grad = None
    loss.backward()
    
    # update
    lr = 0.1 if i < 100000 else 0.01
    for p in parameters:
        p.data += -lr * p.grad

print(f"final loss: {loss.item():.4f}")

final loss: 1.3786


In [10]:
emb = C[Xtr[:1000]]
h = torch.tanh(emb.view(-1, 30) @ W1 + b1)
print(f"h mean: {h.mean().item():.4f}")
print(f"h std: {h.std().item():.4f}")
print(f"saturated: {(h.abs() > 0.97).float().mean().item()*100:.1f}%")

h mean: -0.0513
h std: 0.9664
saturated: 86.2%


86.2% saturated. That means 86% of the neurons are stuck at -1 or 1, and their gradients are basically zero. They're barely learning. WTF!

In [11]:
C  = torch.randn((74, 10))
W1 = torch.randn((30, 200)) * (5/3) / (30**0.5)
b1 = torch.randn(200)
W2 = torch.randn((200, 74))
b2 = torch.randn(74)
parameters = [C, W1, b1, W2, b2]
for p in parameters:
    p.requires_grad = True

In [12]:
for i in range(200000):
    
    # minibatch
    ix = torch.randint(0, Xtr.shape[0], (32,))
    
    # forward pass
    emb = C[Xtr[ix]]
    h = torch.tanh(emb.view(-1, 30) @ W1 + b1)
    logits = h @ W2 + b2
    loss = F.cross_entropy(logits, Ytr[ix])
    
    # backward pass
    for p in parameters:
        p.grad = None
    loss.backward()
    
    # update
    lr = 0.1 if i < 100000 else 0.01
    for p in parameters:
        p.data += -lr * p.grad

print(f"final loss: {loss.item():.4f}")

final loss: 1.6571


In [13]:
emb = C[Xtr[:1000]]
h = torch.tanh(emb.view(-1, 30) @ W1 + b1)
print(f"h mean: {h.mean().item():.4f}")
print(f"h std: {h.std().item():.4f}")
print(f"saturated: {(h.abs() > 0.97).float().mean().item()*100:.1f}%")

h mean: -0.1431
h std: 0.8721
saturated: 51.4%


### Kaiming Correction: 

